designs/executables.md

```diff
@@ -0,0 +1,118 @@
+     * the toolkit.
+     *
+     * Complete exceptionally with [ExecutableValidationException]
+     */
+    fun validate(path: Path): CompletionStage<Nothing>
+}
+```
+
+Registering a new `ExecutableType` will require a `plugin.xml` entry :
+
+```xml
+<executableType implementation="software.aws.toolkits.jetbrains.core.executables.SamExecutable"/>
+```
+
+## Usage of Executables
+The management, validation, and resolution of executables will be managed in a central location. The `ExecutableManager` will expose
+methods that allow the resolution of an executable path and will return types that express the state of that executable.
+
+### Possible States
+A defined executable may have one of several states:
+* **Resolved** the toolkit has a path to an executable, and it is valid (in the case the `ExecutableType` has specified a validation mechanism)
+* **Unresolved** the toolkit doesn't have a current local path at which the executable resides, and is either unable to automatically resolve
+or the automated resolution has failed.
+* **Invalid** the toolkit has a path to the executable, but the path has been deemed invalid (e.g. incorrect version, wrong permissions etc).
+
+These states are conveyed as a `sealed class`:
+
+```kt
+sealed class ExecutableInstance {
+    class Executable(val executablePath: Path) : ExecutableInstance()
+    class InvalidExecutable(val executablePath: Path, val validationError: String) : ExecutableInstance()
+    class UnresolvedExecutable(val resolutionError: String? = null) : ExecutableInstance()
+}
+```
+
+### Operations
+The `ExecutableManager` will expose two methods that allow an executables path to be determined:
+
+```kt
+interface ExecutableManager {
+    fun getExecutable(type: ExecutableType<*>): CompletionStage<ExecutableInstance>
+    fun getExecutableIfPresent(type: ExecutableType<*>): ExecutableInstance.Executable?
+}
+```
+
+* **getExecutable** will asynchronously resolve the path and validate it (if the `ExecutableType` extends `AutoResolvable` and/or `Validatable`).
+The result will be cached (with a timestamp of the file) and future calls will not require auto-resolution.
+* **getExecutableIfPresent** a non-blocking call that will check if the `ExecutableType` has already been resolved and return it if it has. *nb: no
+validation occurs as part of this call*
+
+### Usage Example
+
+When the executable is required in code (e.g. SAM Invoke) it can be used as follows:
+
+```kt
+val manager = ExecutableManager.getInstance()
+when(val executable = manager.getExecutable<SamExecutable>()) {
+  is ExecutableInstance.Executable -> doSomethingWithExecutable(executable.executablePath)
+  is ExecutableInstance.InvalidExecutable -> surfaceErrorToUserAboutInvalidVersion(exectuable.validationError)
+  is ExecutableInstance.UnresolvedExecutable -> promptCustomerToConfigure(executable.resolutionError)
+}
+```
+
+## Open Question
+* How do path's get manually set?
```

Comment:
Can we get all `ExecutableType` from `ExecutableManager` that are not `AutoResolvable`? We could hook this into the toolkit config system somehow, possibly with another interface that helps describe the ExecutableType to the config UI.